save-demo/src/main/kotlin/com/saveourtool/save/demo/service/GithubDownloadToolService.kt

```diff
@@ -0,0 +1,109 @@
+package com.saveourtool.save.demo.service
+
+import com.saveourtool.save.demo.entities.ReleaseAsset
+import com.saveourtool.save.demo.entities.ReleaseMetadata
+import com.saveourtool.save.demo.storage.ToolKey
+import com.saveourtool.save.demo.storage.ToolStorage
+import com.saveourtool.save.demo.utils.toByteBufferFlux
+
+import io.ktor.client.*
+import io.ktor.client.plugins.contentnegotiation.*
+import io.ktor.client.request.*
+import io.ktor.client.statement.*
+import io.ktor.serialization.kotlinx.json.*
+import io.ktor.utils.io.*
+import org.springframework.stereotype.Service
+import reactor.core.publisher.Flux
+import reactor.kotlin.core.publisher.toMono
+
+import java.nio.ByteBuffer
+import javax.annotation.PostConstruct
+
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.Channel
+import kotlinx.serialization.decodeFromString
+import kotlinx.serialization.json.Json
+
+/**
+ * Service that downloads tools from github when starting
+ */
+@Service
+class GithubDownloadToolService(
+    private val toolStorage: ToolStorage,
+) {
+    private val jsonSerializer = Json { ignoreUnknownKeys = true }
+
+    @Suppress("InjectDispatcher")
+    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO)
+    private val httpClient = httpClient()
+
+    private fun getGithubMetadataUrlByKey(key: ToolKey) = with(key) {
+        val release = if (version == LATEST_VERSION) {
+            version
+        } else {
+            "tags/$version"
+        }
+        "$GITHUB_API_URL/$ownerName/$toolName/releases/$release"
+    }
+
+    private fun getMetadata(key: ToolKey): ReleaseMetadata {
+        val channel: Channel<ReleaseMetadata> = Channel()
+        scope.launch {
+            httpClient.get(getGithubMetadataUrlByKey(key))
+                .bodyAsText()
+                .let {
+                    jsonSerializer.decodeFromString<ReleaseMetadata>(it)
+                }
+                .let {
+                    channel.send(it)
+                    channel.close()
+                }
+        }
+        return runBlocking { channel.receive() }
+    }
+
+    @Suppress("ReactiveStreamsUnusedPublisher")
+    private suspend fun downloadAsset(asset: ReleaseAsset): Flux<ByteBuffer> = httpClient.get {
+        url(asset.downloadUrl)
+        accept(asset.contentType())
+    }
+        .bodyAsChannel()
+        .toByteBufferFlux()
+
+    private fun downloadFromGithubAndUploadToStorage(key: ToolKey) = getMetadata(key).assets
+        .filterNot(ReleaseAsset::isDigest)
+        .first()
+        .let { asset ->
+            scope.launch {
+                toolStorage.upload(key, downloadAsset(asset)).subscribe()
+            }
+        }
+
+    @PostConstruct
+    private fun loadToStorage() = toolStorage.list()
+        .collectList()
+        .flatMapIterable { availableFiles ->
+            supportedTools.filter {
+                it !in availableFiles
+            }
+        }
+        .flatMap { key ->
```

Comment:
Logging may be useful here

Reply:
Will add it

Reply:
Provided logging for `GithubDownloadToolService`